### PR TITLE
ARROW-12675: [C++] CSV parsing report row on which error occurred

### PR DESCRIPTION
**Pull request by n3world**
For serial CSV readers track the absolute row number and report it in errors encountered during parsing or converting.

I did try to get row numbers for the parallel reader but the only way I thought that could work would be to add delimiter counting to the Chunker but that seemed to add more complexity than I wanted to.

### CODE REVIEW

File: python/pyarrow/tests/test_csv.py [hunk 1]
```diff
@@ -974,6 +976,36 @@ def read_csv(self, *args, validate_full=True, **kwargs):
         table.validate(full=validate_full)
         return table
 
+    def test_row_numbers_in_errors(self):
+        """ Row numbers are only correctly counted in serial reads """
+        csv, _ = make_random_csv(4, 100, write_names=False)
```
Comment (lidavidm) [inline python/pyarrow/tests/test_csv.py:981]: Is there a particular reason to only test without the header row? It'd be good to check that that gets handled properly too.

File: cpp/src/arrow/csv/reader.cc [hunk 2]
```diff
@@ -888,9 +899,10 @@ class AsyncThreadedTableReader
                            std::shared_ptr<io::InputStream> input,
                            const ReadOptions& read_options,
                            const ParseOptions& parse_options,
-                           const ConvertOptions& convert_options, Executor* cpu_executor)
+                           const ConvertOptions& convert_options,
+                           Executor* cpu_executor)
       : BaseTableReader(std::move(io_context), input, read_options, parse_options,
-                        convert_options),
+                        convert_options, false),
```
Comment (lidavidm) [inline cpp/src/arrow/csv/reader.cc:905]:
```suggestion
                        convert_options, /*count_rows=*/false),
```

Comment (lidavidm) [reply cpp/src/arrow/csv/reader.cc:905]: (and maybe note that the parallel reader cannot count rows yet)

File: cpp/src/arrow/csv/reader.cc [hunk 3]
```diff
@@ -992,7 +1004,7 @@ Result<std::shared_ptr<TableReader>> MakeTableReader(
         io_context, input, read_options, parse_options, convert_options, cpu_executor);
   } else {
     reader = std::make_shared<SerialTableReader>(io_context, input, read_options,
-                                                 parse_options, convert_options);
+                                                 parse_options, convert_options, true);
```
Comment (lidavidm) [inline cpp/src/arrow/csv/reader.cc:1007]:
```suggestion
                                                 parse_options, convert_options, /*count_rows=*/true);
```

File: cpp/src/arrow/csv/reader.cc [hunk 4]
```diff
@@ -1004,7 +1016,7 @@ Future<std::shared_ptr<StreamingReader>> MakeStreamingReader(
     const ParseOptions& parse_options, const ConvertOptions& convert_options) {
   std::shared_ptr<BaseStreamingReader> reader;
   reader = std::make_shared<SerialStreamingReader>(
-      io_context, cpu_executor, input, read_options, parse_options, convert_options);
+      io_context, cpu_executor, input, read_options, parse_options, convert_options, true);
```
Comment (lidavidm) [inline cpp/src/arrow/csv/reader.cc:1019]:
```suggestion
      io_context, cpu_executor, input, read_options, parse_options, convert_options, /*count_rows=*/true);
```

Comment (lidavidm) [reply cpp/src/arrow/csv/reader.cc:1019]: (will probably need to be reformatted here)

File: cpp/src/arrow/csv/reader.cc [hunk 5]
```diff
@@ -500,6 +510,7 @@ class ReaderMixin {
 
   // Number of columns in the CSV file
   int32_t num_csv_cols_ = -1;
+  int64_t num_rows_;
```
Comment (lidavidm) [inline cpp/src/arrow/csv/reader.cc:513]: nit: maybe something more like `last_row_number_` or `num_seen_rows_` and a comment noting `-1` means we aren't tracking the row count?

File: cpp/src/arrow/csv/parser_test.cc [hunk 6]
```diff
@@ -623,5 +643,48 @@ TEST(BlockParser, QuotedEscape) {
...
+    BlockParser parser(options, -1, 0);
+    ASSERT_NO_FATAL_FAILURE(AssertParseOk(parser, csv));
+    int row = 0;
+    auto status = parser.VisitColumn(
+        0, [row](const uint8_t* data, uint32_t size, bool quoted) mutable -> Status {
+          return ++row == 2 ? Status::Invalid("Bad value") : Status::OK();
+        });
+    ASSERT_RAISES(Invalid, status);
+    ASSERT_NE(std::string::npos, status.message().find("Row 1: Bad value"))
+        << status.message();
```
Comment (lidavidm) [inline cpp/src/arrow/csv/parser_test.cc:659]: You can do something like `EXPECT_RAISES_WITH_MESSAGE_THAT(Invalid, ::testing::HasSubstr("Row 1: Bad value", status)`

File: cpp/src/arrow/csv/parser_test.cc [hunk 7]
```diff
@@ -623,5 +643,48 @@ TEST(BlockParser, QuotedEscape) {
...
+  {
+    BlockParser parser(options, -1, -1);
+    ASSERT_NO_FATAL_FAILURE(AssertParseOk(parser, csv));
+    int row = 0;
+    auto status = parser.VisitColumn(
+        0, [row](const uint8_t* data, uint32_t size, bool quoted) mutable -> Status {
+          return ++row == 3 ? Status::Invalid("Bad value") : Status::OK();
+        });
+    ASSERT_RAISES(Invalid, status);
+    ASSERT_EQ(std::string::npos, status.message().find("Row")) << status.message();
```
Comment (lidavidm) [inline cpp/src/arrow/csv/parser_test.cc:685]: And here you can use `::testing::Not(::testing::HasSubstr())`

**Review by lidavidm — COMMENTED** [review]
Thanks for doing this. I left some comments about the tests. I'm not as familiar with the CSV parser so we'll pull in someone else to take a look there too, but this looks good overall.

File: cpp/src/arrow/csv/parser_test.cc [hunk 8]
```diff
@@ -536,22 +537,42 @@ TEST(BlockParser, QuotesSpecial) {
...
+        << st.message();
+  }
+  // No row number
+  {
+    BlockParser parser(ParseOptions::Defaults(), 2 /* num_cols */, -1);
+    auto csv = MakeCSVData({"a\n"});
+    Status st = Parse(parser, csv, &out_size);
+    ASSERT_RAISES(Invalid, st);
+    ASSERT_NE(std::string::npos,
+              st.message().find("CSV parse error: Expected 2 columns, got 1: a"))
```
Comment (lidavidm) [inline cpp/src/arrow/csv/parser_test.cc:574]: Sorry, just a nit, you can use gmock in this block of tests too.

**Review by lidavidm — COMMENTED** [review]
Just another nit about the tests.

File: python/pyarrow/tests/test_csv.py [hunk 9]
```diff
@@ -974,6 +976,56 @@ def read_csv(self, *args, validate_full=True, **kwargs):
...
+
+        # Test without skip_rows and column names in the csv
+        csv_bad_columns = csv + b"1,2\r\n"
+        with pytest.raises(pa.ArrowInvalid,
+                           match="Row 102: Expected 4 columns, got 2"):
+            self.read_bytes(csv_bad_columns, read_options=read_options,
+                            convert_options=convert_options)
+
+        csv_bad_type = csv + b"a,b,c,d\r\n"
+        message = ("In CSV column #0: Row 102: " +
```
Comment (pitrou) [inline python/pyarrow/tests/test_csv.py:996]: This should probably be "Row #102", no?

File: python/pyarrow/tests/test_csv.py [hunk 10]
```diff
@@ -52,18 +52,20 @@ def generate_col_names():
...
+        csv.write(",".join(col_names))
     csv.write(linesep)
     for row in arr.T:
         csv.write(",".join(map(str, row)))
         csv.write(linesep)
     csv = csv.getvalue().encode()
     columns = [pa.array(a, type=pa.int64()) for a in arr]
-    expected = pa.Table.from_arrays(columns, col_names)
+    expected = pa.Table.from_arrays(
+        columns, col_names) if write_names else None
```
Comment (pitrou) [inline python/pyarrow/tests/test_csv.py:68]: I'm not sure what the condition is for here?

File: cpp/src/arrow/csv/reader.cc [hunk 11]
```diff
@@ -319,11 +319,12 @@ class ReaderMixin {
  public:
   ReaderMixin(io::IOContext io_context, std::shared_ptr<io::InputStream> input,
               const ReadOptions& read_options, const ParseOptions& parse_options,
-              const ConvertOptions& convert_options)
+              const ConvertOptions& convert_options, bool count_rows)
       : io_context_(std::move(io_context)),
         read_options_(read_options),
         parse_options_(parse_options),
         convert_options_(convert_options),
+        num_rows_seen_(count_rows ? 1 : -1),
```
Comment (pitrou) [inline cpp/src/arrow/csv/reader.cc:327]: This is weird. Why not have a separate member `bool count_rows_`?

File: cpp/src/arrow/dataset/file_csv.cc [hunk 12]
```diff
@@ -51,7 +51,7 @@ Result<std::unordered_set<std::string>> GetColumnNames(
     const csv::ParseOptions& parse_options, util::string_view first_block,
     MemoryPool* pool) {
   uint32_t parsed_size = 0;
-  csv::BlockParser parser(pool, parse_options, /*num_cols=*/-1,
+  csv::BlockParser parser(pool, parse_options, /*num_cols=*/-1, /*num_rows=*/0,
```
Comment (pitrou) [inline cpp/src/arrow/dataset/file_csv.cc:54]: Are you sure `num_rows` is the argument name?

File: cpp/src/arrow/csv/parser.h [hunk 13]
```diff
@@ -63,19 +63,24 @@ class ARROW_EXPORT DataBatch {
...
       const auto max_pos =
           static_cast<int32_t>(values_buffer->size() / sizeof(ParsedValueDesc)) - 1;
-      for (int32_t pos = col_index; pos < max_pos; pos += num_cols_) {
+      for (int32_t pos = col_index; pos < max_pos; pos += num_cols_, ++row) {
         auto start = values[pos].offset;
         auto stop = values[pos + 1].offset;
         auto quoted = values[pos + 1].quoted;
-        ARROW_RETURN_NOT_OK(visit(parsed_ + start, stop - start, quoted));
+        Status status = visit(parsed_ + start, stop - start, quoted);
+        if (ARROW_PREDICT_FALSE(first_row >= 0 && !status.ok())) {
```
Comment (pitrou) [inline cpp/src/arrow/csv/parser.h:80]: Suggestion:
```c++
if (ARROW_PREDICT_FALSE(!status.ok())) {
  if (first_row >= 0) {
    status = ...
  }
  return status;
}
```

File: cpp/src/arrow/csv/parser.cc [hunk 14]
```diff
@@ -35,14 +35,21 @@ using detail::ParsedValueDesc;
...
 }
 
-Status MismatchingColumns(int32_t expected, int32_t actual) {
-  char s[50];
-  snprintf(s, sizeof(s), "Expected %d columns, got %d", expected, actual);
-  return ParseError(s);
+Status MismatchingColumns(int32_t expected, int32_t actual, int64_t row_num,
+                          util::string_view row) {
+  if (row.length() > 100) {
+    row = row.substr(0, 96);
```
Comment (pitrou) [inline cpp/src/arrow/csv/parser.cc:46]: It probably isn't great to show a truncated row without mentioning that it is truncated.

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks a lot for doing this. Here are some suggestions and questions.

File: cpp/src/arrow/csv/reader.cc (hunk 11 above)
Comment (pitrou) [reply cpp/src/arrow/csv/reader.cc:327]: Having separate member variables would be clearer IMO.

File: cpp/src/arrow/csv/parser.h (hunk 13 above)
Comment (pitrou) [reply cpp/src/arrow/csv/parser.h]: Ah, you're right. But you should still ensure that `!status.ok()` is the first condition inside `ARROW_PREDICT_FALSE`.

File: python/pyarrow/tests/test_csv.py (hunk 10 above)
Comment (n3world) [reply python/pyarrow/tests/test_csv.py]: But then the column names returned by the csv parser may be different than the table or is that not an issue?

Comment (pitrou) [reply python/pyarrow/tests/test_csv.py]: Well, the test should ensure that the column names are the same.

File: cpp/src/arrow/csv/parser.h (hunk 13 above)
Comment (n3world) [reply cpp/src/arrow/csv/parser.h]: I updated it to be almost what you suggested but I kept the ARROW_RETURN_NOT_OK around status. If there is a desire to add an ARROW_RETURN_WITh_CONTEXT macro I am happy to add that and modify the other macros to use it. It only gets rid of a duplicate status.ok() check which isn't much overhead so probably not really worth it.

**Comment by pitrou** [comment]
@ursabot please benchmark

File: cpp/src/arrow/dataset/file_csv.cc [hunk 15]
```diff
@@ -51,7 +51,7 @@ Result<std::unordered_set<std::string>> GetColumnNames(
     const csv::ParseOptions& parse_options, util::string_view first_block,
     MemoryPool* pool) {
   uint32_t parsed_size = 0;
-  csv::BlockParser parser(pool, parse_options, /*num_cols=*/-1,
+  csv::BlockParser parser(pool, parse_options, /*num_cols=*/-1, /*first_row=*/0,
```
Comment (pitrou) [inline cpp/src/arrow/dataset/file_csv.cc:54]: Should this be 1, as is done in the CSV reader?

**Review by pitrou — COMMENTED** [review]
Thanks a lot for the updates! Just a remaining question.

**Review by pitrou — APPROVED** [review]
+1, thank you @n3world

**Pull request closed without merging**